### DISCUSSION TITLE
Generic peripheral overhaul suggestion

### DISCUSSION THREAD

**Discussion by SirEdvin** [cc-tweaked/CC-Tweaked#1093] [General]
So, as modder I am don't like current generic peripheral API for two reasons:
- Implementation is ... strange ... for me
- It closed in un-extendable, so I need to re-implement inventory peripheral over and over, if I want to keep it for peripheral, that I made.

So, I have a suggestion for rework generic peripheral API in more pluggin-like way. I have an [example](https://github.com/SirEdvin/UnlimitedPeripheralWorks/blob/1.18/src/main/kotlin/site/siredvin/peripheralworks/computercraft/ComputerCraftProxy.kt), how it can be.

In short:
- Main peripheral, that can access "plugins", small classes with a lot of lua methods and `additonalType`
- Plugins require specific object for construction
- In the end, main peripheral can have several specific plugins in any case, for example, any item that will pass this pipe and have fluidStorage will have `fluid_storage` plugin attached to it and receive `fluid_storage` methods without any extra implementation

Current design flaws: 
- I don't have solid idea how to insert events in this approach.
